openedx_external_enrollments/external_enrollments/greenfig_external_enrollment.py

```diff
@@ -0,0 +1,113 @@
+"""GreenfigInstanceExternalEnrollment class file."""
+import io
+import logging
+from datetime import datetime
+
+import requests
+from rest_framework import status
+
+from openedx_external_enrollments.edxapp_wrapper.get_site_configuration import configuration_helpers
+from openedx_external_enrollments.edxapp_wrapper.get_student import get_user
+from openedx_external_enrollments.external_enrollments.base_external_enrollment import BaseExternalEnrollment
+from openedx_external_enrollments.models import EnrollmentRequestLog
+
+LOG = logging.getLogger(__name__)
+
+
+class GreenfigInstanceExternalEnrollment(BaseExternalEnrollment):
+    """
+    GreenfigInstanceExternalEnrollment class.
+    """
+
+    def __init__(self):
+        self.DROPBOX_API_URL = configuration_helpers.get_value('DROPBOX_API_URL', 'https://content.dropboxapi.com/2')
+        self.DROPBOX_FILE_PATH = configuration_helpers.get_value('DROPBOX_FILE_PATH', '/courses.txt')
+        self.DROPBOX_TOKEN = configuration_helpers.get_value('DROPBOX_TOKEN', 'token')
+
+    def __str__(self):
+        return 'greenfig'
+
+    def _execute_post(self, url, data=None, headers=None, json_data=None):
+        """
+        Send updated list of courses to dropbox
+        """
+        response = requests.post(
+            url=url,
+            data=json_data,
+            headers=headers,
+        )
+        return response
+
+    def _get_enrollment_headers(self):
+        """Returns headers required to upload a file to dropbox."""
+        return {
+            'Authorization': 'Bearer {token}'.format(token=self.DROPBOX_TOKEN),
+            'Content-Type': 'application/octet-stream',
+            'Dropbox-API-Arg': '{\"path\":\"%s\",\"mode\":{\".tag\":\"overwrite\"}}' % self.DROPBOX_FILE_PATH,
+        }
+
+    def _get_download_headers(self):
+        """Returns headers required to download a dropbox file."""
+        return {
+            'Authorization': 'Bearer {token}'.format(token=self.DROPBOX_TOKEN),
+            'Dropbox-API-Arg': '{\"path\":\"%s\"}' % self.DROPBOX_FILE_PATH,
+        }
+
+    def _get_enrollment_data(self, data, course_settings):
+        """Returns a file in memory with a new or updated enroll."""
+        user, _ = get_user(email=data.get('user_email'))
+        dropbox_file = self._get_course_list(course_settings)
+        dropbox_file = dropbox_file.text.split('\n')
+        enrollment_search_pattern = u'{email}, {course_id}'.format(
+            email=user.email,
+            course_id=course_settings.get('external_course_run_id'),
+        )
+        enrolled = 'true'
+
+        for enrollment in dropbox_file:
+            if enrollment_search_pattern in enrollment:
+                dropbox_file.remove(enrollment)
+                if 'true' in enrollment:
+                    enrolled = 'false'
+
+        enrollment_data = u'{date}, {fullname}, {email}, {course_id}, {enrolled}\n'.format(
+            date=datetime.now().strftime('%m-%d-%Y %H:%M:%S'),
```

Comment:
Could this date format change in the future? @luismorenolopera If so, I think it should be a setting...

Reply:
changed